### PR TITLE
whir: add HVZK-WHIR hiding polynomial commitment scheme

### LINKED ISSUE #1587: [HVZK-WHIR 4/6] HVZK code-switching round
Part of the HVZK-WHIR effort tracked in #1590. Paper: [eprint 2026/391, Section 9.4 / Construction 9.7](https://eprint.iacr.org/2026/391.pdf). Depends on #1584, #1585, #1586.

## What

A WHIR round reduces a claim about a large interleaved code to one about a smaller code. The HVZK version (Construction 9.7) makes two local changes to a round:

1. The prover sends a fresh ZK mask `s = Enc(s̃, r'')` for the smaller code (#1584).
2. The OOD sample is replaced by a `PrivateZeroEvader` (#1585): instead of `y = ze(ρ) · f`, the prover sends `y = ze(ρ) · (f, r, s)` for fresh randomness `r`.

The joint constraint is then carried through the existing batching machinery — no change to that part. Lemmas 9.8 (HVZK simulator) and 9.9 (RBR soundness) cover the analysis.

## Where

`whir/src/pcs/prover.rs` and `verifier.rs`. Add a `zk: bool` flag on the round; the `false` branch keeps today's code unchanged.

## Acceptance criteria

- Per-round prover communication grows by exactly `m_zk · ι_zk + t_ood` field elements over the non-ZK round (matches paper's bound).
- proptest end-to-end on a small instance (e.g., RS over `BinomialExtensionField<KoalaBear, 4>`, `n = 4`, `λ = 20`): honest prover always accepts.
- proptest: simulator composes with #1586's via Theorem 4.5 — the composed simulator's distribution matches the real view.

## Notes

Read **Section 4 (Composition for honest-verifier zero-knowledge IORs)** before starting. The composition theorem (Theorem 4.5) is the structural reason this paper can keep simulators modular; #1586 onward only works if each round's simulator is composable in the paper's sense.

Design choices above are guidelines. If a cleaner decomposition emerges during implementation, raise it in the PR.

### LINKED ISSUE #1588: [HVZK-WHIR 5/6] HVZK base-case IOPP
Part of the HVZK-WHIR effort tracked in #1590. Paper: [eprint 2026/391, Section 7 / Construction 7.2](https://eprint.iacr.org/2026/391.pdf). Depends on #1584, #1586.

## What

When the recursive code-switching reaches its terminal small instance, WHIR runs a non-succinct check. The HVZK base case (Construction 7.2) sends one full extra encoded mask `g = Enc(g̃, r')`, the verifier samples `γ`, and the prover opens `f* = g̃ + γ·f` and a few spot checks. The 2× overhead at the leaf is acceptable because the leaf is tiny (size sublinear in the original witness).

Lemma 7.3 covers HVZK; Lemma 7.4 covers RBR soundness.

## Where

`whir/src/pcs/base_case_zk.rs` (new), with separate `BaseCaseZkProver` / `BaseCaseZkVerifier` types. Reuse the existing base case for the `ZK = false` path.

## Acceptance criteria

- proptest: completeness on small parameters.
- proptest: simulator equivalence (Lemma 7.3) — for RS-based encodings the simulation error is 0.
- proptest: empirical RBR soundness within Lemma 7.4's bounds.

## Notes

Mask-size choice is per Section 2.7's footnote: `Czk` is a Reed–Solomon code with message length `O(λ / log log λ)` and rate `1/log λ`, giving masks of size `Õ(λ)` and `t_zk = O(λ / log log λ)`. Document the choice in code comments so a future reviewer can sanity-check it against the paper.

This is a guideline — the paper's parameter choices assume Reed–Solomon for `Czk`; if the implementation picks a different code via #1584, the parameters need re-deriving and that derivation should be in the PR description.

### LINKED ISSUE #1589: [HVZK-WHIR 6/6] Compose HVZK-WHIR PCS + end-to-end tests
Part of the HVZK-WHIR effort tracked in #1590. Paper: [eprint 2026/391, Section 10 / Theorem 4](https://eprint.iacr.org/2026/391.pdf). Depends on #1586, #1587, #1588.

## What

Wire the building blocks into a `HidingWhirPcs` adapter (mirroring the existing `HidingFriPcs`), and add a `ZK: bool` const generic to `WhirPcs`. The composition follows Theorem 4.5 — no new cryptography here, just plumbing.

## Acceptance criteria

- `HidingWhirPcs<F, EF, …>` implementing `MultilinearPcs<EF, Challenger>`.
- `WhirParameters::new_testing_zk()` and `new_benchmark_zk()` (mirror the FRI helpers in `fri/src/config.rs`).
- `whir/tests/zk_e2e.rs`:
  - 32 random multilinear polynomials, sizes `2^16 ≤ n ≤ 2^20`, full commit/open/verify roundtrip with `ZK = true`.
  - With `ZK = false`, the transcript is byte-identical to today's WHIR (regression guard).
- `whir/benches/zk_overhead.rs`: prover time, verifier time, proof size for `ZK = true` vs `ZK = false`. Paper claims `1 + o(1)` overhead — flag CI regressions.
- Update `whir/src/lib.rs` doc comment to cite both papers (eprint 2024/1586 and eprint 2026/391).

## Out of scope

- R1CS reduction (paper Section 11): orthogonal, separate issue if/when Plonky3 grows R1CS.
- Distance-amplified codes (paper Section 12): orthogonal optimization.
- Holographic / preprocessed verifier: orthogonal.
- Malicious-verifier ZK: implied at the SNARG level via Fiat–Shamir; no extra work needed here.

## Notes

Mostly mechanical once #1586–#1588 land. The interesting bit is making sure the simulator composition in tests actually exercises Theorem 4.5 end-to-end — i.e., the test should construct the composed simulator from the per-round simulators rather than treating the protocol as a monolith.

### LINKED ISSUE #1590: Tracking: HVZK-WHIR — zero-knowledge variant of WHIR (eprint 2026/391)
## Background

WHIR (`whir/`) currently has no zero-knowledge property. The recent paper

> Alessandro Chiesa, Giacomo Fenzi, Guy Weissenberg.
> *Zero-Knowledge IOPPs for Constrained Interleaved Codes.*
> [eprint 2026/391](https://eprint.iacr.org/2026/391.pdf), Feb 2026.

gives an HVZK variant of WHIR with `1 + o(1)` overhead — meaning enabling ZK is essentially free, and the non-ZK path can stay byte-for-byte unchanged. The construction is modular: it adds small "mask" oracles to the existing sumcheck and code-switching rounds rather than redesigning WHIR.

This is a tracking issue for the implementation, decomposed so each sub-issue is one PR reviewable on its own.

## Why now

- ZK is increasingly required for privacy-preserving applications using hash-based SNARKs (anonymous credentials, identity, age assurance — see paper §1).
- The paper's overhead is small enough that ZK can be opt-in at no cost when off.
- The FRI stack already has `HidingFriPcs` as precedent — WHIR is the missing piece.

## Plan

- [x] #1584 — ZK encoding trait + Reed–Solomon (Definition 3.16, Proposition 3.19)
- [x] #1585 — Private zero-evader (Definition 9.2, Lemma 9.3)
- [x] #1586 — HVZK sumcheck with sublinear masks (Section 6, Construction 6.3)
- [ ] #1587 — HVZK code-switching round (Section 9, Construction 9.7)
- [ ] #1588 — HVZK base-case IOPP (Section 7, Construction 7.2)
- [ ] #1589 — Compose HVZK-WHIR PCS + end-to-end tests (Section 10, Theorem 4)

#1584 and #1585 are independent leaves and can be picked up in parallel. #1586 is the meatiest and depends on #1584. #1587 and #1588 each depend on earlier blocks. #1589 is integration plumbing.

## Out of scope (separate issues if/when needed)

- R1CS reduction (paper §11) — Plonky3 is AIR-only today.
- Distance-amplified codes (paper §12) — orthogonal optimization.
- Holographic / preprocessed verifier — orthogonal.
- Malicious-verifier ZK is implied at the SNARG level via Fiat–Shamir; no extra work needed here.

## Notes for contributors

- Each sub-issue carries an API sketch and acceptance criteria. These are guidelines — the paper is precise, but the Rust embedding is a design exercise. If a better factoring emerges during implementation, raise it in the PR.
- Read **Section 4 (Composition for honest-verifier zero-knowledge IORs)** before #1586 onward. The composition theorem (Theorem 4.5) is the structural reason this paper can keep simulators modular; it is what makes the decomposition above possible.
- Each sub-protocol has a simulator test against the real view; that is the right way to gain confidence in the implementation.

### PR DESCRIPTION
**Pull request by tcoratger**
## Summary

Implements the full HVZK-WHIR pipeline of [eprint 2026/391](https://eprint.iacr.org/2026/391) as a hiding multilinear polynomial commitment scheme, composing the previously merged building blocks (#1584 ZK codes, #1585 private zero-evader, #1586 HVZK sumcheck, #1740 code-switch relation).

Closes #1590 and its remaining open sub-issues:
- Closes #1587 (HVZK code-switching round, Construction 9.7)
- Closes #1588 (HVZK base-case IOPP, Construction 7.2)
- Closes #1589 (compose HVZK-WHIR PCS + end-to-end tests, Theorem 4)

## What's added

A parallel hiding pipeline in `whir/src/pcs/zk/` — the plain WHIR path is untouched byte-for-byte:

```text
commit   : interleaved ZK Reed-Solomon encoding of the witness
fold     : masked sumcheck batches        (Construction 6.3)
reduce   : HVZK code-switching rounds     (Construction 9.7)
finish   : non-succinct masked base case  (Construction 7.2)
```

| Module | Content |
|---|---|
| `adapter` | `HidingWhirPcs` implementing `MultilinearPcs`, mirroring `HidingFriPcs` |
| `committer` | interleaved ZK RS commits (per-limb randomness rows), `FoldedRsCode` |
| `constraint/` | symbolic source-claim tracking (Eq/Pow covectors closed under prefix folding) + dense mask claims |
| `code_switch` | Construction 9.7 batching algebra |
| `base_case/` | Construction 7.2 prover/verifier with one-time-pad reveals |
| `prover/`, `verifier` | the round loop on both sides |
| `config`, `proof` | derived ZK budgets and proof payload |

Supporting extensions: aux-claim carry + weight scaling in `p3-sumcheck`'s HVZK overlay, grouped interleaved mask oracles (`stack_codewords`), explicit-randomness draws in `p3-zk-codes`, Fiat-Shamir domain-separator coverage, and a `zk_overhead` criterion bench.

## Design decisions

- **Parallel pipeline, not a `zk` flag**: the carried relation differs (committed-sumcheck relation `<f, W> + sum_i <xi_i, u_i> = target`, Definition 5.8), and this keeps the non-ZK path as a regression guard.
- **Mask grouping**: masks committed together stack into one interleaved oracle (one commitment per sumcheck batch / per code-switch round), so base-case spot checks authenticate a whole group with a single Merkle path per position. This halved the proof-size overhead.
- **No commitment-phase OOD samples**: per the paper's round-by-round analysis, replaced by list-size union bounds.
- **Batching coefficients start at the first challenge power** (`gamma^{i+1}`), so the carried claim keeps coefficient one — paper-faithful.
- **HVZK is tested end-to-end**: a witness-free simulator (honest pipeline on a claim-conditioned random witness) produces an accepting transcript, plus a witness-independence coupling test and tamper-rejection tests asserting exact error variants.

## Measured overhead (KoalaBear, 100-bit security, k = 4, bench params)

| | 2^16 | 2^18 |
|---|---|---|
| Prover time | 1.29x | 1.19x |
| Proof size | 3.28x | 3.04x |

The size overhead is a fixed ~120 KB of mask spot-check openings that amortizes with witness size (the paper's `1 + o(1)` is stated for IOP oracle proof length; after Merkle compilation the additive constant is expected).

## Known follow-ups

- Packed round-0 sumcheck: the initial batch lifts the base-field witness to the extension (unpacked); driving it through the layout-based ZK prover would remove most of the remaining prover-time overhead.
- Single fresh-blind oracle at the base case (common mask domain) would cut the proof-size constant further.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: whir/src/pcs/zk/config.rs [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+    ///
+    /// - The plain WHIR configuration is invalid.
+    /// - The sumcheck mask is shorter than 3 coefficients.
+    /// - An oracle's randomness rows do not fit in its codeword slack.
+    pub fn new(
+        num_variables: usize,
+        params: ProtocolParameters,
+        zk: ZkParameters,
+    ) -> Result<Self, ZkConfigError> {
+        if zk.ell_zk < 3 {
```
Comment (soispoke) [inline whir/src/pcs/zk/config.rs:182]: `ZkParameters` defines security-relevant ZK budgets, but `ZkWhirConfig::new` only
rejects degenerate values. Please either derive these knobs from `security_level`
or validate that caller-provided values meet the ZK security ledger. In
particular, `mask_queries`, `mask_log_inv_rate`, and `ell_zk` need to cover the
mask spot-check / list-size / zero-evader terms, not only the plain WHIR
main-code terms.

File: whir/src/pcs/zk/adapter.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    F: TwoAdicField,
+    EF: ExtensionField<F> + TwoAdicField,
+    Challenger: FieldChallenger<F> + GrindingChallenger<Witness = F>,
+{
+    /// Builds the Fiat-Shamir domain separator for this protocol instance.
+    ///
+    /// Encodes the protocol parameters and the full HVZK transcript shape.
+    /// Every challenge is thereby bound to this configuration.
+    pub fn add_domain_separator<const DIGEST_ELEMS: usize>(&self, ds: &mut DomainSeparator<EF, F>) {
+        ds.commit_statement::<Challenger, DIGEST_ELEMS>(&self.config.inner);
```
Comment (soispoke) [inline whir/src/pcs/zk/adapter.rs]: The ZK path calls the plain commitment domain-separator builder, which still
encodes `add_ood(commitment_ood_samples)`, but the HVZK transcript has no initial
commitment-OOD phase. This isn't a soundness bug by itself, but it's misleading
and easy to misuse. Give the ZK pipeline a dedicated commit separator
shape that omits this phantom OOD step.

File: whir/src/pcs/zk/base_case/tests.rs [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+        err,
+        BaseCaseZkError::MaskSpotCheckFailed {
+            group: 0,
+            position: 2
+        }
+    );
+}
+
+#[test]
+fn base_case_reveals_are_one_time_padded() {
```
Comment (soispoke) [inline whir/src/pcs/zk/base_case/tests.rs:391]: This test is labeled as checking one-time-pad behavior, but it reruns the same
secrets and randomness, so it only checks deterministic replay. Please vary the
hidden message under matched fresh-mask randomness and assert the reveal delta is
`gamma * secret_delta`; otherwise this test would pass even if it did not
meaningfully exercise the OTP property.

**Review by soispoke — COMMENTED** [review]
- ZkWhirConfig` doesn't enforce the ZK security parameters against `security_level` (detail inline on `config.rs`)
- Two smal/non-blockingl issues, a domain-separator cleanup and a hiding test overclaim

**Comment by latifkasuli** [comment]
The main protocol shape looks sound to me: the mask commitments are opened and Merkle-verified, the code-switch OOD answers feed the carried relation instead of being treated as plain evaluations, the base-case `mu_g`/commitments are bound before `gamma`, and the final folded object is handled through the masked base case rather than exposed in clear. I also don't see the old "ZK branch not exercised" issue reappearing in the core path.

A few extra things I noticed, in addition to the inline threads:

1. `MaskCodeShape::new` derives the mask domain in linear `usize` space before the two-adicity guard runs:

```rust
((message_len + randomness_len).next_power_of_two()) << log_inv_rate
```

For very large `mask_log_inv_rate`, this can overflow/panic before `MaskDomainExceedsTwoAdicity` is reachable. This is the same reason the earlier config branch used log-space derivation: compute

```text
log_domain_size = log2(next_power_of_two(message_len + randomness_len)) + log_inv_rate
```

check `log_domain_size <= EF::TWO_ADICITY`, then shift. Small mechanical fix, but it keeps the typed error honest.

2. The test named `zk_whir_end_to_end_multi_round` looks like it only gives one intermediate round with the current fixture:

```text
num_variables = 14
folding = ConstantFromSecondRound(5, 3)
14 - 5 = 9
(9 - 6).div_ceil(3) = 1
```

I'd add an explicit `assert!(config.n_rounds() >= 2)` in that test and bump the fixture if needed, so the round-to-round oracle carry path is actually exercised.

3. Building on @soispoke's note about the base-case one-time-pad test: the PCS-level simulator test (`zk_whir_simulator_produces_accepting_transcript`) has the analogous gap one level up — it runs the honest pipeline on a claim-conditioned witness and checks acceptance, which is a completeness property rather than the composed HVZK/distribution argument (Theorem 4.5). Since this PR also closes #1590, I'd either add a composed-simulator test that compares against a real transcript (byte-equal on the RS/zero-error parts under matched RNG, like the sumcheck-layer `run_view_match_rs` does) or narrow the test name/comments so the Theorem 4.5 claim isn't carried by a completeness test.

4. Minor coverage gap: I don't see an asserting proof-overhead test for the `m_zk * iota_zk + t_ood` accounting, only the benchmark printout. A small assertion would make the #1587 AC easier to audit.

File: whir/src/pcs/zk/config.rs [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    /// # Errors
+    ///
+    /// - The plain WHIR configuration is invalid.
+    /// - The sumcheck mask is shorter than 3 coefficients.
+    /// - An oracle's randomness rows do not fit in its codeword slack.
+    pub fn new(
+        num_variables: usize,
+        params: ProtocolParameters,
+        zk: ZkParameters,
+    ) -> Result<Self, ZkConfigError> {
```
Comment (Nashtare) [inline whir/src/pcs/zk/config.rs:181]: In addition to what's been said, we don't tie it to the targeted security level, we should have a validation layer here to ensure params are satisfactory

File: whir/src/pcs/zk/config.rs (hunk 1 above)
Comment (tcoratger) [reply whir/src/pcs/zk/config.rs:182]: Fixed, `ZkWhirConfig::new` now derives it

File: whir/src/pcs/zk/adapter.rs (hunk 2 above)
Comment (tcoratger) [reply whir/src/pcs/zk/adapter.rs]: Solved with a dedicated ZK commit separator

File: whir/src/pcs/zk/base_case/tests.rs (hunk 3 above)
Comment (tcoratger) [reply whir/src/pcs/zk/base_case/tests.rs:391]: Solved with a real one-time-pad delta test

**Comment by tcoratger** [comment]
> The main protocol shape looks sound to me: the mask commitments are opened and Merkle-verified, the code-switch OOD answers feed the carried relation instead of being treated as plain evaluations, the base-case `mu_g`/commitments are bound before `gamma`, and the final folded object is handled through the masked base case rather than exposed in clear. I also don't see the old "ZK branch not exercised" issue reappearing in the core path.
> 
> A few extra things I noticed, in addition to the inline threads:
> 
> 1. `MaskCodeShape::new` derives the mask domain in linear `usize` space before the two-adicity guard runs:
> 
> ```rust
> ((message_len + randomness_len).next_power_of_two()) << log_inv_rate
> ```
> 
> For very large `mask_log_inv_rate`, this can overflow/panic before `MaskDomainExceedsTwoAdicity` is reachable. This is the same reason the earlier config branch used log-space derivation: compute
> 
> ```
> log_domain_size = log2(next_power_of_two(message_len + randomness_len)) + log_inv_rate
> ```
> 
> check `log_domain_size <= EF::TWO_ADICITY`, then shift. Small mechanical fix, but it keeps the typed error honest.
> 
> 2. The test named `zk_whir_end_to_end_multi_round` looks like it only gives one intermediate round with the current fixture:
> 
> ```
> num_variables = 14
> folding = ConstantFromSecondRound(5, 3)
> 14 - 5 = 9
> (9 - 6).div_ceil(3) = 1
> ```
> 
> I'd add an explicit `assert!(config.n_rounds() >= 2)` in that test and bump the fixture if needed, so the round-to-round oracle carry path is actually exercised.
> 
> 3. Building on @soispoke's note about the base-case one-time-pad test: the PCS-level simulator test (`zk_whir_simulator_produces_accepting_transcript`) has the analogous gap one level up — it runs the honest pipeline on a claim-conditioned witness and checks acceptance, which is a completeness property rather than the composed HVZK/distribution argument (Theorem 4.5). Since this PR also closes [Tracking: HVZK-WHIR — zero-knowledge variant of WHIR (eprint 2026/391) #1590](https://github.com/Plonky3/Plonky3/issues/1590), I'd either add a composed-simulator test that compares against a real transcript (byte-equal on the RS/zero-error parts under matched RNG, like the sumcheck-layer `run_view_match_rs` does) or narrow the test name/comments so the Theorem 4.5 claim isn't carried by a completeness test.
> 4. Minor coverage gap: I don't see an asserting proof-overhead test for the `m_zk * iota_zk + t_ood` accounting, only the benchmark printout. A small assertion would make the [[HVZK-WHIR 4/6] HVZK code-switching round #1587](https://github.com/Plonky3/Plonky3/issues/1587) AC easier to audit.

Thanks a lot @latifkasuli and @soispoke. I fixed all the comments, don't hesitate to let me know if you find something else!

**Review by Nashtare — COMMENTED** [review]
The PR is massive so I'd probably need to go through some of its bits a bit more, but at least I didn't get Codex nor Claude to find any major findings.

File: whir/src/pcs/zk/config.rs [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+        }
+
+        // Derive the mask spot-check count t_zk.
+        //
+        //     reach security on the (1 - delta_zk)^{t_zk} branch
+        //     no PoW on mask spot checks -> target the full security level
+        //     t_zk is also the mask randomness length -> t_zk-query private
+        let security_level = params.security_level;
+        let soundness_type = params.soundness_type;
+        let mask_queries = soundness_type.queries(security_level, zk.mask_log_inv_rate);
```
Comment (Nashtare) [inline whir/src/pcs/zk/config.rs]: Shouldn't it be 
```suggestion
        let union = log2_ceil_usize(2 * n_rounds + 2);
  let mask_queries = soundness_type.queries(security_level + union, zk.mask_log_inv_rate);
```

File: whir/src/pcs/zk/constraint/source.rs [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                    debug_assert!(k <= *num_variables);
+                    // Folding the first k bits scales a power covector by
+                    //
+                    //     prod_{i < k} (1 - g_i + g_i * var^{2^{m-1-i}})
+                    //
+                    // The selection walk squares the seed once per coordinate, in reverse order.
+                    //
+                    //     seed = var^{2^{m-k}}
+                    //     -> factors hit exponents 2^{m-k}, ..., 2^{m-1}
+                    let seed = var.exp_power_of_2(*num_variables - k);
```
Comment (Nashtare) [inline whir/src/pcs/zk/constraint/source.rs]: can't we carry the running seed instead of recomputing? It seems a bit wasteful

File: whir/src/pcs/zk/prover/mod.rs [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+
+            let mask_claim = ZkMaskClaim {
+                base_claim_coeff: EF::ONE,
+                ood_coeffs: ood_coeffs.to_vec(),
+                in_domain_coeffs: query_coeffs.to_vec(),
+            };
+            // Carried scalar: the source residual plus the mask-side total
+            // `sum_i <xi_i, u_i>` at the current covector scales.
+            let carried =
+                batch.residual_prover.claimed_sum() + masks.claims.evaluate(&masks.messages);
```
Comment (Nashtare) [inline whir/src/pcs/zk/prover/mod.rs]: this calls `masks.claims.evaluate(&masks.messages)` twice (here and below) (actually thrice but I'm ignoring the debug assert). May be worth maintaining a running value?

File: whir/src/pcs/zk/prover/mod.rs [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+        prover_data: HidingWhirProverData<F, EF, MT>,
+        claims: &[(Point<EF>, EF)],
+        challenger: &mut Challenger,
+        rng: &mut R,
+    ) -> ZkWhirProof<F, EF, MT> {
+        let config = self.config;
+        let num_variables = config.num_variables;
+        let sumcheck_mask_encoding = config.sumcheck_mask.encoding::<EF>();
+
+        // Claimed evaluations, echoed verbatim into the proof payload.
```
Comment (Nashtare) [inline whir/src/pcs/zk/prover/mod.rs]:
```suggestion
        // Claimed evaluations
```

File: whir/src/pcs/zk/prover/mod.rs [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+            &mut zk_data,
+            &sumcheck_mask_encoding,
+            &self.extension_mmcs,
+            config.folding_factor(0),
+            config.starting_folding_pow_bits,
+            EF::ZERO,
+            challenger,
+            rng,
+        );
+        // Proof accumulators, assembled into the final payload at the end.
```
Comment (Nashtare) [inline whir/src/pcs/zk/prover/mod.rs]:
```suggestion
        // Proof accumulators
```

File: whir/src/pcs/zk/verifier.rs [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+        dims: &[Dimensions],
+        index: usize,
+        opening: &QueryOpening<F, EF, MT::Proof>,
+        round: usize,
+        randomness: &Point<EF>,
+    ) -> Result<EF, ZkVerifierError> {
+        // Defense in depth: bind the opened row length locally.
+        //
+        //     Merkle hashing already pins the committed width,
+        //     but the later fold indexes by 2^folding directly.
```
Comment (Nashtare) [inline whir/src/pcs/zk/verifier.rs]: bit verbose
```suggestion
```

File: whir/src/pcs/zk/config.rs (hunk 5 above)
Comment (tcoratger) [reply whir/src/pcs/zk/config.rs]: Fixed

File: whir/src/pcs/zk/constraint/source.rs (hunk 6 above)
Comment (tcoratger) [reply whir/src/pcs/zk/constraint/source.rs]: Yes fixed

File: whir/src/pcs/zk/prover/mod.rs (hunk 7 above)
Comment (tcoratger) [reply whir/src/pcs/zk/prover/mod.rs]: Fixed

File: whir/src/pcs/zk/config.rs (hunk 4 above)
Comment (tcoratger) [reply whir/src/pcs/zk/config.rs:181]: Fixed

**Review by Nashtare — APPROVED** [review]
Thank you Thomas! This looks really nice! :)
